src/components/layouts/FilterBar.tsx

```diff
@@ -1,60 +1,86 @@
+import { checkAssignee } from '@/utils/assignee'
+import { NoAssigneeExtraOptions } from '@/utils/noAssignee'
+import { getDebouncedFilteredAssignees, setDebouncedFilteredAssignees } from '@/utils/users'
+import { Box, IconButton, Stack } from '@mui/material'
+import { useEffect, useState } from 'react'
+import { useSelector } from 'react-redux'
+import { z } from 'zod'
 
 interface FilterBarProps {
   mode: UserRole
   updateViewModeSetting: (payload: CreateViewSettingsDTO) => void
 }
+
 export const FilterBar = ({ mode, updateViewModeSetting }: FilterBarProps) => {
   const [activeDebounceTimeoutId, setActiveDebounceTimeoutId] = useState<NodeJS.Timeout | null>(null)
   const { view, filteredAssigneeList, filterOptions, assignee, token, viewSettingsTemp } = useSelector(selectTaskBoard)
   const [filteredAssignee, setFilteredAssignee] = useState(filteredAssigneeList)
   const [loading, setLoading] = useState(false)
+  const [initialAssignees, setInitialAssignees] = useState(filteredAssignee)
+  const [latestFetchResults, setLatestFetchResults] = useState(filteredAssignee)
   const viewMode = viewSettingsTemp ? viewSettingsTemp.viewMode : view
   const viewModeFilterOptions = viewSettingsTemp ? (viewSettingsTemp.filterOptions as IFilterOptions) : filterOptions //ViewSettingsTemp used to apply temp values of viewSettings in filterOptions and viewMode because clientSideUpdate applies outdated cached values to original view and filterOptions if navigated
 
   useEffect(() => {
     setFilteredAssignee(filteredAssigneeList)
   }, [filteredAssigneeList]) //to prevent filteredAssignee not updating when filteredAssigneeList changes when fetching assignee is delayed
 
+  useEffect(() => {
+    // Base these initial values off of first fetch of filteredAssignee
+    initialAssignees.length === 0 && setInitialAssignees(filteredAssignee)
+    latestFetchResults.length === 0 && setLatestFetchResults(filteredAssignee)
+  }, [initialAssignees, latestFetchResults, filteredAssignee])
+
+  useEffect(() => {
+    loading && setFilteredAssignee(latestFetchResults)
+  }, [latestFetchResults, loading])
+
+  const handleSelectInputChange = async (newInputValue: string) => {
+    if (newInputValue) {
+      setLoading(true)
+      const newAssignees = await getDebouncedFilteredAssignees(
+        activeDebounceTimeoutId,
+        setActiveDebounceTimeoutId,
+        z.string().parse(token),
+        newInputValue,
+        filterOptions.type,
+      )
+      // Lets an effect handle the latest fetch update instead based on loading state
+      setLatestFetchResults(newAssignees)
+      setLoading(false)
+      return
+    }
+
+    // If the selector is clicked away from, set loading state to false right away
+    setLoading(false)
+    // Reset selector to have initial assignee list. DO NOT use filteredAssignee as it can cause inconsistencies after filteroption is changed
+    setFilteredAssignee(initialAssignees)
+  }
+
```

Comment:
It seems that the `Selector` component for filtering by assignee has accumulated a significant amount of state management and logic. Given its current complexity, it would be beneficial to refactor the entire Selector component, along with its associated logic, into a separate module. Making this change now would be ideal since the implementation is still fresh in our minds, and delaying it could complicate future refactoring efforts as the context might be lost over time. What are your thoughts on proceeding with this refactor?

Reply:
There seems to be a considerable amount of tech debt on the frontend side but refactoring the selector component right now might not be the best approach. We have a ton of priorities for M3, but more importantly we have no idea / confirmation on the new multiple select component. Refactoring could be best done after we have confirmation on that @aatbip

Reply:
I am sorry for the misunderstanding. I didn't mean to write that we refactor the `Selector` component but I meant that we move the filter by assignee `Selector` component implementation from this file to a separate module so that there will be separation of concern. 
Regarding refactoring the `Selector` component, I agree that we should have a better confirmation on it. Thanks for considering. @rrojan

Reply:
@aatbip Ah I see. Solid point!

Still I'm hesitant to refactor this as of right now though... refactoring doesn't make much sense unless we change a lot of code around, and we don't want this to introduce new bugs right now because of it.

Reply:
I completely agree with your point of view. I was suggesting and looking forward to know your thoughts. Thank you for considering. @rrojan